### PR TITLE
feat: add recent searches

### LINKED ISSUE #84: Add Recent Searches
### Description

Add 'recent searches' to the Search Credentials/Badges page. The user should get an overview of recent searches/their search history when accessing the search functionality. Recent searches can be removed.

Screenshot 2023-12-08 at 11 14 11

### Motivation

Improve finding credentials and badges, find previous search results faster.

### Requirements

- [ ] Create "Recent Searches" logic
- [ ] Add functionality to "Me" - Search

### Open Questions

_No response_

### Are you planning to contribute this in a PR?

Yes

### PR DESCRIPTION
**Pull request by Oran-Dan**
# Description of change
We've added the recent searches functionality to the search bar.
A recent search will only be added as an actual credential that was clicked after entering a search term.
This includes being able to delete a recent search.
Clicking on a recent search will move it back to the top of the list.

In the frontend we now use .map instead of .filter to display credentials/connections.
This makes sure the actual order as conveyed by the backend is displayed.

## Links to any relevant issues
#84 

## How the change has been tested
Extensive integration tests located in the unime/src-tauri/tests folder.

## Definition of Done checklist
Add an `x` to the boxes that are relevant to your changes.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by daniel-mader — DISMISSED** [review]

File: identity-wallet/src/state/dev_mode/reducers/ferris_static_profile.rs [hunk 1]
```diff
@@ -191,6 +191,11 @@ pub async fn load_ferris_profile() -> Result<AppState, AppError> {
         },
     ]);
 
+    state.search_results.recent_credentials = vec![
+        DRIVERS_LICENSE_CREDENTIAL.display_credential.id.clone(),
+        OPEN_BADGE.display_credential.id.clone(),
+    ];
```
Comment (nanderstabel) [inline identity-wallet/src/state/dev_mode/reducers/ferris_static_profile.rs:197]: this addition to the state is causing `test_load_dev_profile` and `test_load_dev_profile_twice` to fail right now. 

So probably `four_credentials_redirect_me.json` needs to be updated with these recent credential id's as well :)

File: identity-wallet/src/state/actions.rs [hunk 2]
```diff
@@ -115,7 +117,11 @@ mod bindings {
...
         #[serde(rename = "[User Journey] Cancel")]
         CancelUserJourney,
-        #[serde(rename = "[User Data] Query")]
-        UserDataQuery { payload: UserDataQuery },
+        #[serde(rename = "[Search] Query")]
+        SearchQuery { payload: SearchQuery },
+        #[serde(rename = "[Search] Add Recent")]
+        AddRecentSearch { payload: AddRecentSearch },
+        #[serde(rename = "[Search] Delete Recent")]
+        DeleteRecentSearch { payload: DeleteRecentSearch },
```
Comment (nanderstabel) [inline identity-wallet/src/state/actions.rs]: Minor thing, but for all other actions we only have the first word captalized, so perhaps nice to keep that consistent:
`Add Recent` --> `Add recent`
`Delete Recent` --> `Delete recent`

File: identity-wallet/src/state/search/mod.rs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+#[derive(Default, Serialize, Deserialize, Derivative, TS, Clone, PartialEq, Debug)]
+#[ts(export)]
+#[serde(default)]
+pub struct SearchResults {
+    pub current: Vec<String>,
+    /// Currently we only search credentials, therefore the recents list is called recent_credentials.
+    pub recent_credentials: Vec<String>,
+}
+
+#[typetag::serde(name = "SearchResults")]
```
Comment (nanderstabel) [inline identity-wallet/src/state/search/mod.rs]: To be consistent with the typetag attributes for the other features, please use snake_case here (`search_results`)

File: identity-wallet/src/state/search/mod.rs [hunk 4]
```diff
@@ -0,0 +1,20 @@
+pub mod actions;
+pub mod reducers;
+
+use super::FeatTrait;
+
+use derivative::Derivative;
+use serde::{Deserialize, Serialize};
+use ts_rs::TS;
+
+#[derive(Default, Serialize, Deserialize, Derivative, TS, Clone, PartialEq, Debug)]
```
Comment (nanderstabel) [inline identity-wallet/src/state/search/mod.rs]: The `Derivative` derive can be used to add more fine-grained attributes to struct. However, they are not being utilized here so better to remove `Derivative` from the derive list :+1:

File: identity-wallet/src/state/search/reducers/add_recent_search.rs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+use crate::state::{
+    actions::{listen, Action},
+    AppState,
+};
+
+const MAX_RECENT_SEARCHES: usize = 3;
+
+pub async fn add_recent_search(state: AppState, action: Action) -> Result<AppState, AppError> {
+    if let Some(recent_search) = listen::<AddRecentSearch>(action) {
+        let mut search_results = state.search_results.clone();
```
Comment (nanderstabel) [inline identity-wallet/src/state/search/reducers/add_recent_search.rs]: No need to clone here :) `let mut search_results = state.search_results;` will simply take (mutable) ownership over `state.search_results`.

File: identity-wallet/src/state/search/reducers/delete_recent_search.rs [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+use crate::state::search::SearchResults;
+use crate::state::{
+    actions::{listen, Action},
+    AppState,
+};
+
+pub async fn delete_recent_search(state: AppState, action: Action) -> Result<AppState, AppError> {
+    if let Some(recent_search) = listen::<DeleteRecentSearch>(action) {
+        let search_results: SearchResults = {
+            let mut recent_credentials = state.search_results.recent_credentials.clone();
```
Comment (nanderstabel) [inline identity-wallet/src/state/search/reducers/delete_recent_search.rs]: No need to clone :)

**Review by nanderstabel — CHANGES_REQUESTED** [review]
Just left some minor comments but overall this feature looks great :partying_face: !

File: identity-wallet/src/state/dev_mode/reducers/ferris_static_profile.rs (hunk 1 above)
Comment (daniel-mader) [reply identity-wallet/src/state/dev_mode/reducers/ferris_static_profile.rs:197]: Thanks! Should've run the tests again ... 😅

File: identity-wallet/src/state/actions.rs (hunk 2 above)
Comment (nanderstabel) [reply identity-wallet/src/state/actions.rs]: Same goes actually for:
```rust
#[typetag::serde(name = "[Search] Add Recent")]
impl ActionTrait for AddRecentSearch {
    fn reducers<'a>(&self) -> Vec<Reducer<'a>> {
        vec![reducer!(add_recent_search)]
    }
}
```
and also the frontend needs to reflect the same name change. So basically you can replace all occurrences in the codebase of `[Search] Add Recent` for `[Search] Add recent`

**Review by daniel-mader — DISMISSED** [review]

File: identity-wallet/src/state/actions.rs (hunk 2 above)
Comment (Oran-Dan) [reply identity-wallet/src/state/actions.rs]: should be good to go now

**Review by nanderstabel — APPROVED** [review]

**Review by daniel-mader — APPROVED** [review]

File: identity-wallet/src/state/search/reducers/add_recent_search.rs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    if let Some(recent_search) = listen::<AddRecentSearch>(action) {
+        let mut search_results = state.search_results;
+
+        search_results
+            .recent_credentials
+            .retain(|recent| recent != &recent_search.id);
+        search_results.recent_credentials.insert(0, recent_search.id);
+
+        if search_results.recent_credentials.len() > MAX_RECENT_SEARCHES {
+            search_results.recent_credentials.remove(MAX_RECENT_SEARCHES);
```
Comment (Norlock) [inline identity-wallet/src/state/search/reducers/add_recent_search.rs:20]: This code is fine, but .pop(); will always ensure the last one is removed

Comment (Oran-Dan) [reply identity-wallet/src/state/search/reducers/add_recent_search.rs:20]: good comment! just merged it already, sorry

Comment (Norlock) [reply identity-wallet/src/state/search/reducers/add_recent_search.rs:20]: haha ok no problem last thing I saw:

```
  // makes the <input> element available to the parent component (example usage: conditionally put focus it)
```
On a svelte export let. we can change comment style to /** mak.... */ to work with intellisense but for now its ok :+1:

**Pull request merged**